network_simulator/link.py

```diff
@@ -95,3 +133,133 @@ def __eq__(self, other):
+        """
+        :ivar link: link that must send its next packet
+        :ivar buff_elem: temporary variable containing LinkBufferElement to send
+        """
+        self.link = link
+        self.buff_elem = None
+
+    def run(self, main_event_loop, statistics):
+        """
+        runs by popping off top packet
+        :param MainEventLoop main_event_loop: event loop for global timing
+        :param Statistics statistics: the Statistics to update
+        """
+
+        # sanity check
+        if self.link.queue.empty():
+            logger.warning("LinkSendEvent while Link empty!")
+            return
+
+        # now pop
+        self.buffer_elem = self.link.pop()
+
+        # update statistics only if not RouterPacket
+        if not isinstance(self.buffer_elem.packet, RouterPacket):
+            statistics.link_packet_transmitted(self.link, \
+                    self.buffer_elem.packet, \
+                    main_event_loop.global_cloc_sec)
+
+    def schedule_new_events(self, main_event_loop):
+        """
+        Schedules new Events. This is called immediately after run().
+        :param MainEventLoop main_event_loop: event loop where new Events will
+        be scheduled.
+        """
+        # schedule a *ReceivedPacketEvent with propagation + transmission
+        # delay, depending on whether Host or Router
+        if isinstance(self.buffer_elem.dest_dev, Router):
+            main_event_loop.schedule_event_with_delay(\
+                    RouterReceivedPacketEvent(0),\
+                    self.link.static_delay_sec + self.packet.size_bits / \
+                        float(self.link.capacity_bps))
+                # TODO(Laksh): RouterReceivedPacketEvent signature
+                # recommended signature RouterReceivedPacketEvent(router, packet)?
+        else:
+            main_event_loop.schedule_event_with_delay(\
+                    HostReceivedPacketEvent(0),\
+                    self.link.static_delay_sec + self.packet.size_bits / \
+                        float(self.link.capacity_bps))
+
+        if self.link.link_buffer.queue.empty():
+            # if link_buffer.queue is empty, then link is no longer busy
+            self.link.busy = False
+        else:
+            # queue new event with just transmission delay
+            assert(self.link.busy == True)
+            main_event_loop.schedule_event_with_delay(LinkSendEvent(self.link),\
+                    self.buffer_elem.packet.size_bits / float(self.link.capacity_bps))
+
+
+class DeviceToLinkEvent(Event):
+    """
+    Event representing a packet going from Device to Link's buffer
+    """
+
+    def __init__(self, packet, link, dest_dev):
+        """
+        :ivar packet: packet arriving at Link
+        :ivar link: link that packet is arriving at
+        :ivar dest_dev: destination Device of the packet pushed onto link
+        """
+        self.packet = packet
+        self.link = link
+        self.dest_dev = dest_dev
```

Comment:
Add assertion that either link.end_1_device == dest_dev or link.end_2_device == dest_dev, just as a check for now.